admin-password-updater/main.go

```diff
@@ -0,0 +1,153 @@
+)
+
+func main() {
+	var managementURI, caFile string
+	u := &updater.PasswordUpdater{}
+
+	flag.StringVar(
+		&u.DefaultUserFile,
+		"default-user-file",
+		"/etc/rabbitmq/conf.d/11-default_user.conf",
+		"Absolute path to file containing default user username and (updated) password. "+
+			"Its directory will be watched for changes.")
+	flag.StringVar(
+		&u.AdminFile,
+		"admin-file",
+		"/var/lib/rabbitmq/.rabbitmqadmin.conf",
+		"Absolute path to file used by rabbitmqadmin CLI. "+
+			"It contains RabbitMQ admin username (must be the same as default user username) and (old) password.")
+	flag.StringVar(
+		&managementURI,
+		"management-uri",
+		"http://127.0.0.1:15672",
+		"RabbitMQ Management URI")
+	flag.StringVar(
+		&caFile,
+		"ca-file",
+		"/etc/rabbitmq-tls/ca.crt",
+		"This file contains the trusted certificate for RabbitMQ server authentication.")
+	klog.InitFlags(nil)
+	flag.Parse()
+	log := klogr.New().WithName("password-updater")
+	u.Log = log
+
+	rabbitClient, err := newRabbitClient(log, managementURI, caFile)
+	if err != nil {
+		log.Error(err, "failed to create RabbitMQ client")
+		return
+	}
+	u.Rmqc = rabbitClient
+
+	// Watch the directory because the file gets re-created (i.e. first removed and then created)
+	// when password is updated by Vault agent and fsnotify will stop watching a re-created file.
+	watchDir := filepath.Dir(u.DefaultUserFile)
+	u.WatchDir = watchDir
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Error(err, "failed to create watcher")
+		return
+	}
+	defer watcher.Close()
+	u.Watcher = watcher
+
+	// Remove trailing new line (.rabbitmqadmin.conf has only one section).
+	ini.PrettySection = false
+
+	// This channel will contain a value when the Pod gets terminated.
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGTERM, syscall.SIGINT)
+
+	// This channel will contain a value when our program terminates itself.
+	// This is preferred over calling os.Exit() because os.Exit() does not run deferred functions.
+	done := make(chan bool, 1)
+	u.Done = done
+
+	go u.HandleEvents()
+
+	log.V(1).Info("start watching", "directory", watchDir)
+	if err := watcher.Add(watchDir); err != nil {
+		log.Error(err, "cannot watch", "directory", watchDir)
+		return
+	}
+
+	select {
+	case sig := <-sigs:
+		log.V(1).Info("terminating", "signal", sig.String())
+	case <-done:
+		log.V(1).Info("terminating")
+	}
+}
+
+func newRabbitClient(log logr.Logger, managementURI, caFile string) (updater.RabbitClient, error) {
+	if strings.HasPrefix(managementURI, "https") {
+		caCert, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			log.Error(err, "failed to read CA file", "file", caFile)
+			return nil, err
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tlsConfig := &tls.Config{
+			RootCAs: caCertPool,
+		}
+		transport := &http.Transport{TLSClientConfig: tlsConfig}
+		rmqc, err := rabbithole.NewTLSClient(managementURI, "", "", transport)
```

Comment:
Interested what the effect is on management API peer verification - I think this would mean peer verification on the management API isn't supported with this flow, though it's not widely used at all so that may be OK. Could be worth calling out.

Reply:
That's a good call.

When a user sets
```
management.ssl.verify = verify_peer
management.ssl.fail_if_no_peer_cert = true
```
I'd expect that RabbitMQ HTTP API clients are supposed to be configured only with client certificates?
In that use case, why would the admin user have a password in the first place?
If the admin user doesn't have a password, there is no need to rotate the password, hence no need to use the admin-password-updater.

As of today, i.e. before this PR, the cluster-operator will always create username and password putting it into a K8s secret. The whole flow of an admin user set up with client certs instead of username and password isn't well supported and not documented at all in the cluster-operator.  It's something we discussed with @yaronp68 and @cdickmann on Friday. We may want to support it in future. Therefore, I wouldn't call it out specifically for the Vault integration.